File: src/effectScope.ts
import { endTrack, startTrack } from './internal.js';
import { Link, Subscriber, SubscriberFlags } from './system.js';

export let activeEffectScope: EffectScope | undefined = undefined;

export function untrackScope<T>(fn: () => T): T {
	const prevSub = activeEffectScope;
	setActiveScope(undefined);
	try {
		return fn();
	} finally {
		setActiveScope(prevSub);
	}
}

export function setActiveScope(sub: EffectScope | undefined): void {
	activeEffectScope = sub;
}

export function effectScope(): EffectScope {
	return new EffectScope();
}

export class EffectScope implements Subscriber {
	// Subscriber
	deps: Link | undefined = undefined;
	depsTail: Link | undefined = undefined;
	flags: SubscriberFlags = SubscriberFlags.None;

	run<T>(fn: () => T): T {
		const prevSub = activeEffectScope;
		setActiveScope(this);
		try {
			return fn();
		} finally {
			setActiveScope(prevSub);
		}
	}

	stop(): void {
		startTrack(this);
		endTrack(this);
	}
}

File: kotlin/EffectScope.kt
package alien.signals

class EffectScope : Subscriber {

    private var deps: MutableList<Signal<*>> = mutableListOf()
    private var flags: Int = 0

    override val deps: MutableList<Signal<*>>
        get() = deps

    fun run(fn: () -> Unit) {
        val prevSub = activeSub
        val prevTrackId = activeTrackId
        setActiveSub(this, nextTrackId())
        startTrack(this)
        try {
            fn()
        } finally {
            setActiveSub(prevSub, prevTrackId)
            endTrack(this)
        }
    }

    fun stop() {
        startTrack(this)
        endTrack(this)
    }
}
